File: src/Wrappers/AntRunner.Wrapper.Ruby/lib/AntRunner.rb
require "json"

module AntRunner
	module Interface
		class Ant			
			def Name
				"No Name"
			end

			def Action
				@Action
			end

			def Action= v
				@Action = v
			end
			
			def Initialize(mapWidth, mapHeight, color, startX, startY)
				#Initialize code
			end

			def Tick(state)
				#Tick code
			end

			def self.Descendants()
				ObjectSpace.each_object(::Class).select {|klass| klass < self }
			end
		end

		module AntAction
			Wait = 0
			MoveRight = 1
			MoveDown = 2
			MoveLeft = 3
			MoveUp = 4
			EchoRight = 5
			EchoDown = 6
			EchoLeft = 7
			EchoUp = 8
			ShieldOn = 9
			ShieldOff = 10
			DropBomb = 11
			ShootRight = 12
			ShootDown = 13
			ShootLeft = 14
			ShootUp = 15
		end

		module Item
			Empty = 0
			SteelWall = 1
			BrickWall = 2
			Bomb = 3
			PowerUpBomb = 4
			PowerUpHealth = 5
			PowerUpShield = 6
			RedAnt = 7
			BlueAnt = 8
			GreenAnt = 9
			OrangeAnt = 10
			PinkAnt = 11
			YellowAnt = 12
			GrayAnt = 13
			WhiteAnt = 14
			RedHome = 15
			BlueHome = 16
			GreenHome = 17
			OrangeHome = 18
			PinkHome = 19
			YellowHome = 20
			GrayHome = 21
			WhiteHome = 22
			Flag = 23
		end

		module GameEvent
			Nothing = 0
			CollisionDamage = 1
			ImpactDamageRight = 2
			ImpactDamageDown = 4
			ImpactDamageLeft = 8
			ImpactDamageUp = 16
			ShotDamageRight = 32
			ShotDamageDown = 64
			ShotDamageLeft = 128
			ShotDamageUp = 256
			BombDamage = 512
			PickUpBomb = 1024
			PickUpShield = 2048
			PickUpHealth = 4096
			PickUpFlag = 8192
			Dead = 16384
			GameOver = 32768
		end

		module ItemColor
			None = 0
			Red = 1
			Blue = 2
			Green = 3
			Orange = 4
			Pink = 5
			Yellow = 6
			Gray = 7
			White = 8
		end

		module ItemBonusValues
			Health = 25
			Shield = 25
			Bomb = 4
		end

		module DamageValues
			Collision = 5
			Impact = 10
			Shot = 20
			Bomb = 30
		end

		class EchoResponse
			attr_accessor :Distance, :Item

			def initialize(distance, item)
				@Distance = distance
				@Item = item
			end
			
			def self.from_json(string)
				data = JSON.load(string)
				return self.new data['Distance'], data['Item']
			end
		end

		class GameState
			attr_accessor :TickNumber, :HasFlag, :FlagX, :FlagY, :AntWithFlag, :Response, :Event

			def initialize(tickNumber, event = GameEvent::Nothing, response = nil, flagX = -1, flagY = -1, antWithFlag = ItemColor::None)
				@TickNumber = tickNumber
				@Event = event
				@Response = response
				@HasFlag = flagX != -1
				@FlagX = flagX
				@FlagY = flagY
				@AntWithFlag = antWithFlag
			end

			def self.from_json(string)
				data = JSON.load(string)
				return self.new(data['TickNumber'], data['Event'], data['Response'].nil? ? nil : EchoResponse.new(data['Response']['Distance'], data['Response']['Item']), data['FlagX'], data['FlagY'], data['AntWithFlag'])
			end
		end		
	end
end

File: src/Wrappers/AntRunner.Wrapper.Php/lib/AntRunner.php
<?php namespace AntRunner_Interface;

class Ant
{
	public $Name;
	public $Action;
	
	public function Initialize($mapWidth, $mapHeight, $antColor, $startX, $startY) {}
	
	public function Tick(GameState $state) {}
}

abstract class AntAction
{
    const Wait = 0;
    const MoveRight = 1;
	const MoveDown = 2;
	const MoveLeft = 3;
	const MoveUp = 4;
	const EchoRight = 5;
	const EchoDown = 6;
	const EchoLeft = 7;
	const EchoUp = 8;
	const ShieldOn = 9;
	const ShieldOff = 10;
	const DropBomb = 11;
	const ShootRight = 12;
	const ShootDown = 13;
	const ShootLeft = 14;
	const ShootUp = 15;
}

abstract class DamageValues
{
    const Collision = 5;
    const Impact = 10;
	const Shot = 20;
	const Bomb = 30;
}

abstract class GameEvent
{
    const Nothing = 0;
    const CollisionDamage = 1;
	const ImpactDamageRight = 2;
	const ImpactDamageDown = 4;
	const ImpactDamageLeft = 8;
	const ImpactDamageUp = 16;
	const ShotDamageRight = 32;
	const ShotDamageDown = 64;
	const ShotDamageLeft = 128;
	const ShotDamageUp = 256;
	const BombDamage = 512;
	const PickUpBomb = 1024;
	const PickUpShield = 2048;
	const PickUpHealth = 4096;
	const PickUpFlag = 8192;
	const Dead = 16384;
	const GameOver = 32768;
}

abstract class Item
{
    const Empty = 0;
    const SteelWall = 1;
	const BrickWall = 2;
	const Bomb = 3;
	const PowerUpBomb = 4;
	const PowerUpHealth = 5;
	const PowerUpShield = 6;
	const RedAnt = 7;
	const BlueAnt = 8;
	const GreenAnt = 9;
	const OrangeAnt = 10;
	const PinkAnt = 11;
	const YellowAnt = 12;
	const GrayAnt = 13;
	const WhiteAnt = 14;
	const RedHome = 15;
	const BlueHome = 16;
	const GreenHome = 17;
	const OrangeHome = 18;
	const PinkHome = 19;
	const YellowHome = 20;
	const GrayHome = 21;
	const WhiteHome = 22;
	const Flag = 23;
}

abstract class ItemBonusValues
{
    const Health = 25;
    const Shield = 25;
	const Bomb = 4;
}

abstract class ItemColor
{
    const None = 0;
    const Red = 1;
	const Blue = 2;
	const Green = 3;
	const Orange = 4;
	const Pink = 5;
	const Yellow = 6;
	const Gray = 7;
	const White = 8;
}

class EchoResponse
{
	public $Distance;
	public $Item;

	function __construct($distance, $item)
	{
		$this->Distance = $distance;
		$this->Item = $item;
	}
}

class GameState
{
	public $TickNumber;
	public $HasFlag;
	public $FlagX;
	public $FlagY;
	public $AntWithFlag;
	public $Response;
	public $Event;
		
	function __construct($tickNumber, $event = GameEvent::Nothing, $response = null, $flagX = -1, $flagY = -1, $antWithFlag = ItemColor::None)
	{
		$this->TickNumber = $tickNumber;
		$this->HasFlag = $flagX != -1;
		$this->FlagX = $flagX;
		$this->FlagY = $flagY;
		$this->AntWithFlag = $antWithFlag;
		$this->Response = $response;
		$this->Event = $event;
	}
}